File: src/taxoquiz/game/seed.py
"""Shareable game seeds.

A seed identifies one game: the same seed always gives the same secret animal,
so two people can play the same round independently. Daily mode is the same
mechanism with the seed derived from the date rather than chosen at random.

Format is `FFFF-BBBBBB`, e.g. `3KQ7-2M9XPT`:

- `FFFF` fingerprints the **dataset**. The example ships 530 species and a full
  scrape has 18,421, so the same body would otherwise mean a different animal
  depending on who you sent it to — and silently, which is the worst version.
  A mismatch is rejected with a message saying so.
- `BBBBBB` is the body: random for practice, derived from the date for daily.

The alphabet excludes I, L, O and U, so a seed can be read aloud or retyped
without 1/I or 0/O confusion.

Seeds are deliberately *not* secret. The mapping is a plain hash of a public
species list, so someone who wants to work out their own seed's answer from the
source can. That is a fair trade for seeds being short, offline-checkable and
requiring no server state; there is nothing to protect here beyond making the
answer non-obvious at a glance.
"""

import hashlib
import secrets
from datetime import date

# Crockford-style: no I, L, O, U.
ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
FINGERPRINT_LEN = 4
BODY_LEN = 6


def _encode(digest: bytes, length: int) -> str:
    n = int.from_bytes(digest, "big")
    out = []
    for _ in range(length):
        n, rem = divmod(n, len(ALPHABET))
        out.append(ALPHABET[rem])
    return "".join(reversed(out))


def fingerprint(species: list[dict]) -> str:
    """A short, stable tag for the dataset's species list.

    Derived from the names in tree order, so any change to the dataset — a
    different scrape, a different `MIN_SITELINKS`, a re-extraction that drops
    species — produces a different tag and old seeds stop resolving instead of
    quietly pointing at something else.
    """
    h = hashlib.sha256()
    for s in species:
        h.update(s["common_name"].encode())
        h.update(b"\n")
    return _encode(h.digest(), FINGERPRINT_LEN)


def _body_for_date(day: date) -> str:
    return _encode(hashlib.sha256(f"taxoquiz-daily:{day.isoformat()}".encode()).digest(), BODY_LEN)


def new_body() -> str:
    """A random seed body. `secrets` rather than `random` so concurrent games
    on the same process cannot land on the same seed."""
    return "".join(secrets.choice(ALPHABET) for _ in range(BODY_LEN))


def make_seed(species: list[dict], body: str | None = None, *, day: date | None = None) -> str:
    """Build a full seed. Pass `day` for the daily seed, `body` to rebuild a known one."""
    if day is not None:
        body = _body_for_date(day)
    return f"{fingerprint(species)}-{body or new_body()}"


def normalise(seed: str) -> str:
    """Accept what a person actually types: any case, spaces, missing dash."""
    cleaned = "".join(c for c in seed.upper() if c in ALPHABET)
    if len(cleaned) != FINGERPRINT_LEN + BODY_LEN:
        raise ValueError(
            f"'{seed}' is not a seed — expected {FINGERPRINT_LEN + BODY_LEN} "
            f"characters like ABCD-234567."
        )
    return f"{cleaned[:FINGERPRINT_LEN]}-{cleaned[FINGERPRINT_LEN:]}"


def resolve(seed: str, species: list[dict]) -> dict:
    """Return the species a seed names, for this dataset.

    Raises ValueError if the seed was made for a different dataset, rather than
    returning a different animal for the same seed.
    """
    seed = normalise(seed)
    fp, body = seed.split("-")
    expected = fingerprint(species)
    if fp != expected:
        raise ValueError(
            f"Seed '{seed}' is for a different dataset (its tag is {fp}, "
            f"this one is {expected}). Both players need the same dataset."
        )
    index = int.from_bytes(hashlib.sha256(f"taxoquiz-seed:{body}".encode()).digest(), "big")
    return species[index % len(species)]

File: frontend/src/engine/seed.ts
/**
 * Shareable game seeds. Port of `taxoquiz/game/seed.py`, which explains the
 * format and why a seed fingerprints its dataset.
 *
 * This is the module where a near miss costs most. A seed shared from the
 * website must open the same animal on a phone, and the two now compute it
 * independently — so the hashing is byte-for-byte Python's, including treating
 * the whole 256-bit digest as one integer, which is what `BigInt` is for.
 */
import { sha256 } from './sha256'
import type { RawNode } from './taxonomy'

export const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ'
export const FINGERPRINT_LEN = 4
export const BODY_LEN = 6

const utf8 = new TextEncoder()

function digestInt(text: string): bigint {
  let n = 0n
  for (const byte of sha256(utf8.encode(text))) n = (n << 8n) | BigInt(byte)
  return n
}

function encode(n: bigint, length: number): string {
  const base = BigInt(ALPHABET.length)
  const out: string[] = []
  for (let i = 0; i < length; i++) {
    out.push(ALPHABET[Number(n % base)])
    n /= base
  }
  return out.reverse().join('')
}

export function fingerprint(species: RawNode[]): string {
  return encode(digestInt(species.map((s) => `${s.common_name}\n`).join('')), FINGERPRINT_LEN)
}

/** `isoDay` is `YYYY-MM-DD`. */
export function bodyForDate(isoDay: string): string {
  return encode(digestInt(`taxoquiz-daily:${isoDay}`), BODY_LEN)
}

/** Today in UTC — the same date `App.tsx` stamps a saved session with, so a
 *  daily round and the check that expires it agree on when the day ends. */
export function utcToday(): string {
  return new Date().toISOString().slice(0, 10)
}

export function newBody(): string {
  // 256 is a multiple of 32, so a byte modulo the alphabet is unbiased.
  const bytes = crypto.getRandomValues(new Uint8Array(BODY_LEN))
  return Array.from(bytes, (b) => ALPHABET[b % ALPHABET.length]).join('')
}

export function makeSeed(species: RawNode[], opts: { body?: string; day?: string } = {}): string {
  const body = opts.day !== undefined ? bodyForDate(opts.day) : opts.body
  return `${fingerprint(species)}-${body || newBody()}`
}

/** Accept what a person actually types: any case, spaces, missing dash. */
export function normalise(seed: string): string {
  const cleaned = [...seed.toUpperCase()].filter((c) => ALPHABET.includes(c)).join('')
  if (cleaned.length !== FINGERPRINT_LEN + BODY_LEN) {
    throw new Error(
      `'${seed}' is not a seed — expected ${FINGERPRINT_LEN + BODY_LEN} characters like ABCD-234567.`,
    )
  }
  return `${cleaned.slice(0, FINGERPRINT_LEN)}-${cleaned.slice(FINGERPRINT_LEN)}`
}

export function resolve(seed: string, species: RawNode[]): RawNode {
  seed = normalise(seed)
  const [fp, body] = seed.split('-')
  const expected = fingerprint(species)
  if (fp !== expected) {
    throw new Error(
      `Seed '${seed}' is for a different dataset (its tag is ${fp}, this one is ${expected}). ` +
      'Both players need the same dataset.',
    )
  }
  return species[Number(digestInt(`taxoquiz-seed:${body}`) % BigInt(species.length))]
}
